### PR TITLE
Ensure that tasks wait for running indirect setup

### PR DESCRIPTION
**Pull request by dstandish**
*Note*: this doesn't change existing trigger rule but does extract the inner functions into methods. For ease of review i have done this part in the first commit, separate from the introduction of setup constraint.

Alternative to https://github.com/apache/airflow/pull/33570

This is yet another approach to solving this one.  But I feel it's the best one thus far because it does not make any changes to existing trigger rule logic, apart from pulling out some internal functions into methods.

The indirect setup constraint logic is processed prior to trigger rules and is completely separate.

This _should_ play nicely with all mapping scenarios but more testing is required to verify.

There are a couple choices to make here.

One is in the "setup constraint" logic, should we apply it only to indirect setups or to all relevant setups?  I think the only time it makes a difference is when it there is a mixture of skipped and failed upstream setups and how we determine state of the object task.

Currently I only look at indirect setups.

Another thing is, for calculation of state, do we just go based on the state of the first non-succcess?  Or should we wait until everything is done and calculate at that point?  Currently it just takes the first non-success and fails (or skips) fast.  And I think this is consistent with airflow's many-to-one behavior more generally.

Note: for easier review, the first commit is a no-behavior-change refactor that just pulls the inner funcions out into methods.  The second commit is where the setup constraint logic is added.

Here is an example dag that can be used to test the behavior:
[Uploading setup_teardown_bug.txt…]()

The way to use that dag is to first change all the "scenarios" to `["success", "success", "success"]` and then you'll get a clean run.
Then you change the scenario e.g. to `["skip", "fail", "succes"]`.  Then, you clear the task with indirect upstream setup and see what happens.  The first mapped setup will skip, 5 seconds later the next fails, then 5 seconds later there's a success.

### CODE REVIEW

File: airflow/ti_deps/deps/trigger_rule_dep.py
Comment (hussein-awala) [inline airflow/ti_deps/deps/trigger_rule_dep.py]: I don't think we have None values in passed:
```suggestion
                    if not status.passed and changed:
```

File: airflow/ti_deps/deps/trigger_rule_dep.py
Comment (hussein-awala) [inline airflow/ti_deps/deps/trigger_rule_dep.py]: Do we really need an expensive group by here?
```suggestion
            upstream = session.execute(
                select(func.count("*"))
                .where(TaskInstance.dag_id == ti.dag_id, TaskInstance.run_id == ti.run_id)
                .where(
                    or_(
                        *self._iter_upstream_conditions(
                            ti=ti, relevant_tasks=indirect_setups, session=session
                        )
                    )
                )
            ).scalar_one()
```

File: airflow/ti_deps/deps/trigger_rule_dep.py
Comment (hussein-awala) [inline airflow/ti_deps/deps/trigger_rule_dep.py]:
```suggestion
            elif done and skipped:
                new_state = TaskInstanceState.SKIPPED
```
With
```
done = upstream == (success + skipped + failed + upstream_failed + removed)
```

File: airflow/ti_deps/deps/trigger_rule_dep.py
Comment (hussein-awala) [inline airflow/ti_deps/deps/trigger_rule_dep.py]: I never understood why why we should have at least one success upstream and the task should not be mapped to change its state to removed.
Also IMHO we need a condition on `done`, otherwise it will not be deterministic.

File: airflow/ti_deps/deps/trigger_rule_dep.py
Comment (hussein-awala) [inline airflow/ti_deps/deps/trigger_rule_dep.py]: Oups, I was breaking my little feature in my PR! :+1:

File: airflow/ti_deps/deps/trigger_rule_dep.py
Comment (hussein-awala) [inline airflow/ti_deps/deps/trigger_rule_dep.py]: I think we can replace all of this by `done` after adding it:
```suggestion
        if not done:
```

**Review by hussein-awala — APPROVED** [review]
It looks good overall; however, we can optimize some of the used queries. I will create a new PR for 2.7.2 once this one is merged.

File: airflow/ti_deps/deps/trigger_rule_dep.py
Comment (dstandish) [reply airflow/ti_deps/deps/trigger_rule_dep.py]: i would rather leave this to future.

File: airflow/ti_deps/deps/trigger_rule_dep.py
Comment (dstandish) [reply airflow/ti_deps/deps/trigger_rule_dep.py]: do we need to wait for done if we already have one that is skipped?

Comment (hussein-awala) [reply airflow/ti_deps/deps/trigger_rule_dep.py]: yes, waiting for a failed setup, in this case we will always have an upstream_failed state if we have at least one failed setup. Without this condition we will have a skipped state once a setup task is skipped, even if it will be followed by a failed setup.

File: airflow/ti_deps/deps/trigger_rule_dep.py
Comment (dstandish) [reply airflow/ti_deps/deps/trigger_rule_dep.py]: > Also IMHO we need a condition on done, otherwise it will not be deterministic.

yeah, the reason why i felt it is ok is that airflow is already a race when there are many-to-one.  so this seems basically consistent with the existing behavior -- to fail or skip fast.  i think in practice it is not likely to make much difference. but...

for others, the scenario is, if there are multiple upstream indirect setups, should we wait for all of them to complete before determining the state of the object task (and then e.g. if there is a combination of success / skip / fail, then we fail it) or should we just take the state of the first non-success?

taking the state of the first non-success is how things work currently when there's a many-to-one direct upstream e.g. `[a, b] >> c`.  in this way it "fails fast" (or skips).  but indirect upstream setup is of course something a little different.  we could choose to wait until all upstream setups are done before marking state.  this means we do *not* fail fast, and the payoff is there is no race condition w.r.t. state.

cc @jedcunningham @ephraimbuddy @potiuk

Comment (hussein-awala) [reply airflow/ti_deps/deps/trigger_rule_dep.py]: The main problem here is that the skipped state is used as one of the successful states (along with success), most of the users do not consider it as failure, it simply means that there is no need to run the task, nor to send a notification (on_failure).

And the problem will be much complicated if we discuss scenarios where there is an external task sensor waits for this task...

I'd love to hear from others what they think.

Comment (dstandish) [reply airflow/ti_deps/deps/trigger_rule_dep.py]: so with your example, if the upstream setups are skipped, the task will be skipped. it's only if there's a mixed state, which is actually a pretty odd situation, don't think super common generally, but we must to decide what to do.
the question is take the first one or wait for all and the evaluate.
if your position is "wait for all and evaluate" what would be your logic?
```
if at least one failure, then upstream_failed
else if at least one skip, skip
```
?
i think it's probably fine... add an implicit all_done for the setups and then evaluate.  i don't think it makes a _huge_ difference either way.

Comment (jedcunningham) [reply airflow/ti_deps/deps/trigger_rule_dep.py]: I lean more toward consistency here (fast fail), even thought that consistency is a non-deterministic outcome.

**Review by jedcunningham — APPROVED** [review]

File: airflow/ti_deps/deps/trigger_rule_dep.py
Comment (dstandish) [reply airflow/ti_deps/deps/trigger_rule_dep.py]: this could be true but, this logic was added to deal with a deadlock and i don't really want to mess with it here. i'm just keeping the behavior consistent with trigger rule. maybe let's do this in followup?

File: airflow/ti_deps/deps/trigger_rule_dep.py
Comment (uranusjr) [reply airflow/ti_deps/deps/trigger_rule_dep.py]: Any reason this is resolved without a comment?

Comment (dstandish) [reply airflow/ti_deps/deps/trigger_rule_dep.py]: I accepted the suggestion but then mypy complained so I reverted it

Comment (dstandish) [reply airflow/ti_deps/deps/trigger_rule_dep.py]: Fundamentally there's nothing wrong with saying is false when that's really what you're looking for

Comment (uranusjr) [reply airflow/ti_deps/deps/trigger_rule_dep.py]: Cool

Comment (dstandish) [reply airflow/ti_deps/deps/trigger_rule_dep.py]: i guess mypy mighta been unrelated

File: airflow/ti_deps/deps/trigger_rule_dep.py
Comment (uranusjr) [inline airflow/ti_deps/deps/trigger_rule_dep.py]: I wonder if we can actually cache this globally; a trigger rule is accessed repeatedly in the schedule in every loop iteration, and the result of this function may change depending on upstream states. We could add the cache decorator because the function was function-local previously, but now it’s on the class I’m not so sure.

File: airflow/ti_deps/deps/trigger_rule_dep.py
Comment (uranusjr) [inline airflow/ti_deps/deps/trigger_rule_dep.py]: And this cache has no effect at all since it is local to this function and only accessed exactly once during its lifetime.

File: airflow/ti_deps/deps/trigger_rule_dep.py
Comment (dstandish) [reply airflow/ti_deps/deps/trigger_rule_dep.py]: ok so i guess we can enclose it again...

Comment (dstandish) [reply airflow/ti_deps/deps/trigger_rule_dep.py]: well, i'll look into this tomorrow i guess

File: airflow/ti_deps/deps/trigger_rule_dep.py
Comment (uranusjr) [inline airflow/ti_deps/deps/trigger_rule_dep.py]: Probably want to move this above `_get_relevant_upstream_map_indexes` (or that function below this). There’s not much difference but better for readability and to prevent accidental access to this variable before it is created.

(Why do we want this dict to be empty when task is teadrown though? Or is there no difference?)

Comment (dstandish) [reply airflow/ti_deps/deps/trigger_rule_dep.py]: see line 554, where we set the dict equal to `relevant_setups` when processing setup constraint.  then at 566 (if we get there) we reassign it to direct upstream tasks

Comment (dstandish) [reply airflow/ti_deps/deps/trigger_rule_dep.py]: so, when `_get_relevant_upstream_map_indexes` is called for setup constraint, then it has access to the setup tasks, and when it is called for direct upstreams, it will have access to the direct upstreams.

we could avoid all of this by passing the task instead of the id, but for whatever reason it feels icky to me to do that since the hash functions are so weird, even if there's no practical way the object gets mutated during processing.  i can be easily convinced though.

Comment (uranusjr) [reply airflow/ti_deps/deps/trigger_rule_dep.py]: But `relevant_setups` is just a subset of all upstreams, isn’t it? Because I don’t see we catch KeyError anywhere, just passing the entire task dict shouldn’t change the logic at all from what I can tell. Is it just for performance?

Comment (dstandish) [reply airflow/ti_deps/deps/trigger_rule_dep.py]: no, cus the bug we're solving here, is the case of _indirect_ setups -- setups that are not direct upstream relatives.  in normal operation this is a non issue.  but when you clear a work task, it clears its setups, though there might be other tasks between the setup and the work task (thus indirect); if you don't add a check for the indirect setups then they will run concurrently. that's the problem here.

Comment (dstandish) [reply airflow/ti_deps/deps/trigger_rule_dep.py]: separately, just a note, i expect there's a way to optimize this and reduce queries from 2 back down to 1; but i don't want to "change multiple things at once" in this way; i just want to layer the setup constraint on top of the existing trigger rule, and apply it only when dag has setups, to reduce potential for impact on dags not using the new feature, particularly given that here we're just trying to introduce a bugfix in a timely manner.

Comment (uranusjr) [reply airflow/ti_deps/deps/trigger_rule_dep.py]: No I mean, why can’t we just put all the tasks in that dict? Maybe `task_dict` should just be `ti.task.dag.task_dict`?

File: airflow/ti_deps/deps/trigger_rule_dep.py
Comment (uranusjr) [inline airflow/ti_deps/deps/trigger_rule_dep.py]: I feel these two should be put back to near the top of the function, or at least before the (nested) function that uses them.

File: airflow/ti_deps/deps/trigger_rule_dep.py
Comment (dstandish) [reply airflow/ti_deps/deps/trigger_rule_dep.py]: yeah true, that should be fine.

File: airflow/ti_deps/deps/trigger_rule_dep.py
Comment (dstandish) [reply airflow/ti_deps/deps/trigger_rule_dep.py]: yeah true

**Comment by dstandish** [comment]
have to update a couple tests since now we can't test the two constraints independently. i'll do tomorrow.

**Review by uranusjr — APPROVED** [review]

**Comment by ephraimbuddy** [comment]
I would prefer that we move this to 2.7.2 considering that it touches the heart of Airflow and we are cutting the RC tomorrow. What do you think?

**Comment by potiuk** [comment]
I'd take the risk to be honest @ephraimbuddy. And maybe @hussein-awala might chime-in, but I think this bug was undermining rather important use case of setup/teardown and it also changes a behaviour from 2.7.0 that someone might start relying on. So this is really a bugfix, but for some people it might break their workflows (literally).

If we also keep the old behaviour in 2.7.1 and fix it in 2.7.2, it will be a bit harder to explain that this is a bugfix. Also 2.7.0 had quite a few teething problems and people went back to 2.6.3 and eagerly waiting for 2.7.1 (I hope) so likely not many people had a chance to start relying in 2.7.0 behaviour, so taking a bit of risk now might be worth it.

**Comment by hussein-awala** [comment]
I understand that `+602 −236 ` is scary for a patch version, but most of the changed code is documentation and tests, and for the changes in `trigger_rule_dep`, @dstandish has encapsulated the processing logic in a method (added an indent) but he changed nothing in the processing behavior for dags that do not use the new feature.

I think the risk level is very low here, especially after all the tests I ran with @dstandish.

**Comment by ephraimbuddy** [comment]
Cherrypicked!

**Pull request merged**
